### PR TITLE
more annotations for aws xray sdk

### PR DESCRIPTION
**Pull request by Sam-Martin**
Adding some previously missing annotations (mostly return types) for the aws-xray-sdk's `Recorder` class.
Puzzling this out based on the original code here: [aws_xray_sdk/core/recorder.py](https://github.com/aws/aws-xray-sdk-python/blob/master/aws_xray_sdk/core/recorder.py)

Notable oddities:

1. `record_subsegment` has a return of `Any` because it returns the value of a wrapped method whose return is unknowable (or if we can do it with generics I don't know how in Python!)
2. `dynamic_naming` has a value of `Any | DefaultDynamicNaming` due to the issue I asked [a question about here](https://github.com/python/typing/discussions/861).
3. `get_trace_entity`/`set_trace_entity` correction from `Context` to `Segment | Subsegment | DummySegment | DummySubsegment`. 
This was an error in my original Pull Request populating aws-xray-sdk initially. 
The `DummySegment` I added because it is set as a segment on `self.context` by [`begin_segment `](https://github.com/aws/aws-xray-sdk-python/blob/master/aws_xray_sdk/core/recorder.py#L237) and therefore can be returned by `get_trace_entity`, therefore should be permissible to _set_. I then added `DummySubsegment` for symmetry's sake.

### CODE REVIEW

File: stubs/aws-xray-sdk/aws_xray_sdk/core/recorder.pyi [hunk 1]
```diff
@@ -50,24 +50,24 @@ class AWSXRayRecorder:
...
+    def begin_subsegment(self, name: str, namespace: str = ...) -> DummySubsegment | Subsegment | None: ...
+    def current_subsegment(self) -> Subsegment | None: ...
     def end_subsegment(self, end_time: time.struct_time | None = ...) -> None: ...
     def put_annotation(self, key: str, value: Any) -> None: ...
     def put_metadata(self, key: str, value: Any, namespace: str = ...) -> None: ...
-    def is_sampled(self): ...
-    def get_trace_entity(self): ...
+    def is_sampled(self) -> bool: ...
+    def get_trace_entity(self) -> Segment | Subsegment: ...
     def set_trace_entity(self, trace_entity: Context) -> None: ...
```
Comment (Akuli) [inline stubs/aws-xray-sdk/aws_xray_sdk/core/recorder.pyi:67]: Are these really supposed to have different types?

**Review by Akuli — COMMENTED** [review]
Just something I noticed while skimming:

File: stubs/aws-xray-sdk/aws_xray_sdk/core/recorder.pyi (hunk 1 above)
Comment (Sam-Martin) [reply stubs/aws-xray-sdk/aws_xray_sdk/core/recorder.pyi]: Good spot! Fixed.

**Review by srittau — APPROVED** [review]
LGTM (without comparing with the implementation)

**Pull request merged**